### PR TITLE
Move test control logic to Haskell

### PR DESCRIPTION
**Pull request by martijnbastiaan**
This PR moves test control from TCL to Haskell. Moving this to Haskell makes sure that we can easily control the individual steps (programming, test control) and interleave it with other steps, such as starting GDB servers, and monitoring UART output. We mostly do this to make sure we can move to a software based development loop.

# TODO
- [X] Write wrapper to control Vivado from Haskell (@martijnbastiaan)
  - [x] ~~Introduce `VivadoM` for ease of usage (@martijnbastiaan)?~~
- [X] Rewrite existing logic in Shake / HardwareTest.tcl (@JvWesterveld)
  - [x] Fix failing HITL tests (rebasing first might help)
  - [X] Fix clock control report generation in `bittide-tools/clockcontrol/plot/Main.hs`, which is broken for the new HITL test definition types (initial attempt in f35d240)
- [x] Rebase on staging branch (see #595)
- [x] Wait for https://github.com/bittide/bittide-hardware/actions/runs/10743081981
- [x] Wait for https://github.com/bittide/bittide-hardware/actions/runs/10792200155

# Future work
  - [ ] Cleanup VIO and ILA probe detection and usage code in `bittide-shake/src/Clash/Shake/Vivado.hs`
  - [ ] Perhaps let `execPrint` from `vivado-hs` print Vivado output to stdout directly instead of only when the relevant Tcl code has finished executing
  - [ ] Perhaps remove most functions in `vivado-hs/src/Vivado/Tcl.hs` and instead use `execPrint` or `execCmd` directly
  - [ ] Perhaps replace `System.Clock` with `Data.Time(.Clock)` and `Data.Time.Format` in `bittide-shake/src/Clash/Shake/Vivado.hs` to print fancier formatted test durations

### CODE REVIEW

**Review by martijnbastiaan — COMMENTED** [review]
Overall this LGTM, great work @JvWesterveld!

I'm only missing some documentation in a couple of places, and I think `Iter` is a misnomer.

File: .github/synthesis/all.json [hunk 1]
```diff
@@ -12,17 +12,17 @@
...
-  {"top": "fullMeshSwCcTest",              "stage": "test",   "targets": "All"          },
-  {"top": "hwCcTopologyTest",              "stage": "test",   "targets": "All"          },
-  {"top": "linkConfigurationTest",         "stage": "test",   "targets": "All"          },
-  {"top": "syncInSyncOut",                 "stage": "test",   "targets": "All"          },
-  {"top": "temperatureMonitor",            "stage": "test",   "targets": "All"          },
-  {"top": "transceiversUpTest",            "stage": "test",   "targets": "All"          },
-  {"top": "vexRiscvTest",                  "stage": "test",   "targets": "Specific [-1]"}
+  {"top": "boardTestExtended",             "stage": "test"},
+  {"top": "boardTestSimple",               "stage": "test"},
+  {"top": "extraProbesTest",               "stage": "test"},
```
Comment (JvWesterveld) [inline .github/synthesis/all.json:17]:
```suggestion
```
`extraProbesTest` was removed (as it tested the now removed YAML configuration files), so CI should fail on this. I think it's a good idea to run all tests (not just those in `staging.json`) once more before merging.

File: .github/synthesis/staging.json [hunk 2]
```diff
@@ -1,8 +1,7 @@
 [
-  {"top": "fullMeshHwCcTest",              "stage": "test", "targets": "All"},
-  {"top": "fullMeshSwCcTest",              "stage": "test", "targets": "All"},
-  {"top": "linkConfigurationTest",         "stage": "test", "targets": "All"},
-  {"top": "safeDffSynchronizer",           "stage": "hdl"},
```
Comment (JvWesterveld) [inline .github/synthesis/staging.json:5]: `safeDffSynchronizer` didn't cause problems with this PR, right?

**Review by JvWesterveld — COMMENTED** [review]
Latest changes LGTM. I have a few more mostly minor comments, partially concerning my own code.

File: .github/synthesis/staging.json (hunk 2 above)
Comment (martijnbastiaan) [reply .github/synthesis/staging.json:5]: Nope :). Accidental delete!

**Review by JvWesterveld — APPROVED** [review]

**Pull request merged**